File: 191.位-1-的个数.go
/*
 * @lc app=leetcode.cn id=191 lang=golang
 *
 * [191] 位1的个数
 */

// @lc code=start
func hammingWeight(num uint32) int {
	res := 0
	for num != 0 {
		num = num & (num - 1) // 取并的结果是将这个数的二进制最后一位不为 0 的位给置 0。
		res++
	}
	return res
}

// @lc code=end

File: 191.位-1-的个数.java
/*
 * @lc app=leetcode.cn id=191 lang=java
 *
 * [191] 位1的个数
 */

// @lc code=start
public class Solution {
    // you need to treat n as an unsigned value
    public int hammingWeight(int n) {
        int tmp = n;
        int count = 0;

        // solution 1
        while (tmp != 0) {
            tmp = tmp & (tmp - 1);
            count++;
        }

        // solution 2
        // while (tmp != 0) {
        //     if ((tmp & 1) == 1) {
        //         count++;
        //     }
        //     tmp = tmp >>> 1;
        // }

        return count;
    }
}
// @lc code=end

